Introduce GMD + ATD

thagikura opened the pull request:
- Introduce Gradle Managed Device definition in a single module (core-database at the moment) using Automated Test Device

dturner commented:
GMD setup is failing with: 

```
Execution failed for task ':core:database:pixel4api30Setup'.
> A failure occurred while executing com.android.build.gradle.internal.tasks.ManagedDeviceSetupTask$ManagedDeviceSetupRunnable
   > java.lang.IllegalStateException: Snapshot setup ran successfully, but the snapshot failed to be created. This is
     likely to a lack of disk space for the snapshot. Try the cleanManagedDevices
     task with the --unused-only flag to remove any unused devices for this project.
```

https://github.com/android/nowinandroid/actions/runs/3673045398/jobs/6212504226

thagikura commented:
I ran it 10 times and failed it due to the lack of disk space 2 times.
https://github.com/android/nowinandroid/actions/runs/3673045398

I'll see if it can be stabilized.

thagikura commented:
After this commit https://github.com/android/nowinandroid/pull/479/commits/392408cfa29374f89fed499fe0681a78de852352 as advised in https://issuetracker.google.com/issues/216744701, 

2 attempts still failed out of 10 attempts.

Now I enabled more logging in this set of Actions.
https://github.com/android/nowinandroid/actions/runs/3683976750

If we still see failed jobs, I'm considering  filing a bug against GMD.

thagikura commented:
The last 10 attempts succeeded as in https://github.com/android/nowinandroid/actions/runs/3683976750

Do you think it's acceptable @dturner ?
In case if it fails in the future, I also enabled the detailed logging to pass it to the product team.

Also this PR only enables GMD in the core-database module, the rest of the modules need to be addressed, too.

dturner commented on core/database/build.gradle.kts:
```diff
@@ -34,6 +34,18 @@ android {
     }
     namespace = "com.google.samples.apps.nowinandroid.core.database"
+
+    testOptions {
```
Can we move this into a common plugin so that GMD is used for every module?

thagikura replied:
There was a bug where defining GMD wasn't possible in Kotlin that made impossible to define it as a convention plugin. (b/242515648)

That is now fixed, but is included in Flamingo canary 9.

alexvanyo replied:
I was going to suggest the same, thanks for the info. Can we create a TODO and a GitHub issue to track?

dturner reviewed (changes requested):
Let's try using GMD everywhere so we can compare it with ReactiveCircus for the whole codebase.

thagikura commented:
> Let's try using GMD everywhere so we can compare it with ReactiveCircus for the whole codebase.

Agreed. Let's see the flakiness to compare with ReactiveCircus.

thagikura commented:
> Agreed. Let's see the flakiness to compare with ReactiveCircus.

Added GMD to modules that have instrumented tests. Can you have a look again?

alexvanyo commented on .github/workflows/AndroidCIWithGmd.yaml:
```diff
@@ -0,0 +1,34 @@
+    steps:
+    - uses: actions/setup-java@v3
+      with:
+        distribution: 'temurin'
```
I don't know if it matters, but what is the reason for `temurin` here? I think in our other workflows we have `zulu`

thagikura replied:
I don't have a strong opinion here either. Let's stick with other workflow files. Changed this to `zulu`.

alexvanyo commented on .github/workflows/AndroidCIWithGmd.yaml:
```diff
@@ -0,0 +1,34 @@
+        java-version: '11'
+    - uses: actions/checkout@v2
+
+    - name: Setup Android SDK
+      uses: android-actions/setup-android@v2
```
Is this necessary? I don't see it in the other build scripts we have.

thagikura replied:
Thanks. It wasn't necessary. The last attempt succeeded without this plugin.

alexvanyo reviewed (approved):
Woo! 🎉 

<img width="824" alt="Screen Shot 2023-01-04 at 2 29 53 PM" src="https://user-images.githubusercontent.com/4217560/210662250-fe057086-407e-4287-870d-467135bb7901.png">

Looks good to me, just two nits in the workflow and it's unfortunate we can't make a convention plugin at the moment. Ideally we could apply that plugin everywhere, so we can't make a mistake of leaving out any new tests by not adding the managed devices configuration.

thagikura commented:
Thanks Alex. Addressed your feedback. 
Filed #523 as a reminder to convert GMD definitions as a convention plugin.
@dturner is it possible to have a look again and resolve the change request if it looks good?

Thanks!

alexvanyo commented on app/build.gradle.kts:
```diff
@@ -73,6 +73,7 @@ android {
         unitTests {
             isIncludeAndroidResources = true
         }
+        // TODO: Convert it as a convention plugin once Flamingo goes out (b/242515648)
```
Nit: Let's avoid links to bugs that aren't publicly accessible, we could link to the newly created GitHub issue instead.

thagikura replied:
Done. Sgtm!

thagikura commented:
Friendly reminder to @dturner please resolve the change request if the PR looks good. Thanks!

dturner reviewed (approved):
Many thanks for doing this.
